### ISSUE TITLE
Incorrect values passed to scale should at least raise a warning

### ISSUE THREAD

**Issue by vaughnkoch** [vega/vega#1528]
Vega reports parse errors well. However, it seems to silently fail displaying marks when it can't process some of the data. It would be very helpful when debugging to be able to find out exactly what's failing.

For example, here's a few simple scales and a line graph. The color input is obviously false, but Vega doesn't seem to report it - it just doesn't show the line. If Vega can't display a mark for some reason, it should at least warn instead of silently dropping it.

```
"scales": [
  {
    "name": "color",
    "type": "ordinal",
    "range": ['#4949ff', '#c13535'],  // Blue, red
    "domain": [0, 1],
  },
  {
    "name": "x",
    "domain": [0, 1],
    "range": "width",
  },
  {
    "name": "y",
    "domain": [0, 1],
    "range": "height",
  },
],

"marks": [
  {
    "type": "line",
    "from": {"data": "density"},
    "encode": {
      "enter": {
        "x": {"scale": "x", "field": "x"},
        "y": {"scale": "y", "field": "y"},
        "stroke": {"scale": "color", "value": "COMPLETELY_BAD_VALUE_HERE" },
        "strokeWidth": {"value": 2}
      },
      "update": {
        "interpolate": {"value": "cardinal"},
        "fillOpacity": {"value": 1}
      },
      "hover": {
        "fillOpacity": {"value": 0.5}
      }
    }
  },
```

Also, if there's a setting I've missed which would trap these situations, I would love to know about it.

**Comment by g3o2** [comment]
Maybe this issue could further specify the term "incorrect". 

My guess is that what you mainly refer to is type errors. Such values are invalid or incompatible, no matter how you define the domain of the scale.

However, I am not sure if it is useful to warn you each time a theoretically valid value is off-domain. Off-domain could also mean that you're not interested in seeing the said value.

**Comment by g3o2** [comment]
For ordinal scales, also take a look at the [`domainImplicit`](https://vega.github.io/vega/docs/scales/#ordinal) property, the [impute](https://vega.github.io/vega/docs/transforms/impute/) transform and [production rules](https://vega.github.io/vega/docs/marks/). Those can help you guard against undefined values.

For quantitative scales, out of range values can be dealt with using the `scales.clamp` property.

**Comment by jheer** [comment]
Yes, this is difficult to do for a number of reasons. In some cases unseen (i.e., out of domain) values may be legal and even wanted, as @g3o2 notes. In other cases, "missing" values such as null or undefined values might pass through a scale but one doesn't want to raise errors for every such occurrence. We do, however, already warn if a log scale domain includes non-positive numbers.

That said, I'm happy to consider possible modifications given more precise criteria for how the envisioned functionality would work!

**Comment by vaughnkoch** [comment]
One of the frustrating parts about building Vega specs is how easy they are to get wrong with simple typos. These errors come to mind:

- syntax errors - these are properly reported which is great
- Not following the value convention, e.g. `"y": -25` (fails) instead of `"y": {"value": -25 }` (works)
- Type errors as mentioned above, passing "foo" to a scale that expects a number 
- Option not valid for the existing directive, e.g. tickCount for scales - I can't remember whether Vega reports these correctly or silently ignores them.

One idea I had would be to add some kind of debug/developer mode which would throw warnings or raise exceptions when Vega sees one of the above behaviors. My experience with Vega has been that when the chart doesn't do what you expect it to do, there's a lot of trial and error involved in figuring out what the problem is.

**Comment by hydrosquall** [comment]
Adding runtime checks about data type validation would make execution of the scales slower, but I agree it may be helpful as a separate mode that can be opted into. 

The vega logger has different severity levels, we could either make use of that or report errors more aggressively (maybe "strict mode" vs "regular" in vega-view's runtime). 

It could check situations where a scaling function returns a value that won't graph properly, like "null" or "Infinity" in a quantitative scale, or it could attempt to validate inputs to the scaling function.
